### ISSUE TITLE
Calculate solar intensity / insolation for a segment/run/ski area

### ISSUE THREAD

**Issue by dhimmel** [dhimmel/openskistats#28]
Calculating the amount of sunlight (insolation) that hits a run segment would be a powerful way for us to quantify the effect of trail orientation. I think it's most helpful to start thinking here in terms of run segments, because I imagine we could then aggregate any metric to the level of a whole run, ski area, or ski area aggregation.

For the MVP, I could see us incorporating:

- sun's declination, where latitude is the input
- segment bearing and slope

This would assume no vegetative shadowing, cloud cover, and limited topographic shadowing. Up for discussion.

Let's use this issue to discuss methods and implementations. @joshimmel is interested in methods here. @nbarbier-s2 has opened a POC in https://github.com/dhimmel/openskistats/pull/24.

**Comment by joshimmel** [comment]
Agreed, it's best to start simple and ignore shadowing effects, which would require the use of continuous raster data (digital elevation maps + canopy height / digital surface maps) beyond the ski area of interest.

To begin, it appears the Python photovoltaic library [pvlib](https://pvlib-python.readthedocs.io/en/stable/) possesses the functionalities we are interested in.

Using position on earth (latitude/longitude), trail segment aspect, and trail segment slope, and assuming either a single day/hour (winter solstice @ solar noon), we can produce solar irradiance values (W/m²) for each trail segment, then aggregate. Obviously limited, but I think a necessary step to infer capacity for melt dependent on "which way you ski".

A couple of later considerations / improvements:

- [ ] Integrate across winter season (approximately Dec 21–Mar 20) as opposed to a single time
- [ ] Incorporate cloud cover percentage or cloud radiative effect (extracted from a global product like [CERES SYN1deg](https://asdc.larc.nasa.gov/project/CERES/CER_SYN1deg-1Hour_Terra-Aqua-MODIS_Edition4A)


Happy to discuss/work on an MVP with @nbarbier-s2 which ideally capitalizes on the existing outputs from the project.

**Comment by dhimmel** [comment]
[sunpath3d](https://andrewmarsh.com/apps/staging/sunpath3d.html) is helpful for building mental models of the sun's path by latitude and date.

> assuming a single day/hour (winter solstice @ solar noon)

A single day / hour is a nice simplifying assumption for a first pass. Do we ultimately expect integrating over all daylight hours across a typical ski season will produce proportionally similar results, or could it change the ranking of ski areas?

> we can produce solar irradiance values (W/m²) for each trail segment, then aggregate

Presumably we want to scale or weight somehow. For example, if adding a trail to a ski area, it should not increase its solar irradiance unless the trail receives more sun than the average of the remaining ski area. What do we weight by? Vertical, 3d distance, etc?

We have observed an eastward orientation preference across the totality of ski areas, presumably because morning sun exposure is preferable when temperatures are cooler. If we were to eventually model this, it seems to be going beyond just solar irradiance by including temperature to model melt-inducing irradiance. Is there terminology for this phenomenon?

**Comment by dhimmel** [comment]
[Here](https://chatgpt.com/share/676edc44-4fe8-8011-b186-497acd3c7380) is what ChatGPT o1 suggested for:

> Using Python's pvlib, calculate solar irradiance at a given location (lat, lon, ele) and a given slope/bearing to represent the orientation of a ski run segment. For now, just compute this at solar noon on the winter solstice based on the location.

**Comment by joshimmel** [comment]
> [Here](https://chatgpt.com/share/676edc44-4fe8-8011-b186-497acd3c7380) is what ChatGPT o1 suggested for:

Using a similar ChatGPT [output](https://chatgpt.com/c/676c4116-8024-8011-bc3e-9de43db8965c) I began yesterday and modifying it to represent Dartmouth Skiway, on the solstice, for the NbNW aspects and a slope of 20, here is a plot of irradiance through the day (peaking at 60 W/m2). No built in functionality for irradiance on SunPath3d. Will look into this more, but as a gut check I feel these values are in the right ballpark. 
![image](https://github.com/user-attachments/assets/993a596e-5ec8-47d4-a761-a036fb02b1e6)

**Comment by joshimmel** [comment]
Plotting daily irradiance sums (Wh/m2/day) across an arbitrarily defined ski season (Dec 01 2024 to Apr 15, 2025), than summing those, we can produce total irradiance for any lat/long aspect slope combo. Perhaps we preserve the daily irradiance and produce daily normalized sums for each ski area. Then we can produce an animation or interactive graphic/table of solar irradiance through the season (which may vary more radically for mountains with narrower aspect distributions). 

![image](https://github.com/user-attachments/assets/70bb985f-3cd7-4702-88fd-10a75611d27f)

To do: 
- [ ] Integrate pvlib analysis into openskistats 
- [ ] Decide whether to incorporate cloudiness (reflected in GHI (Global Horizontal Irradiance))
- [ ] Determine how to normalize daily irradiance values (3d distance?)
- [ ] Determine whether aggregating these normalized values most accurately reflects melt potential

**Comment by dhimmel** [comment]
We've made good progress here by:

1. computing solar irradiation throughout the course of a typical season for run segments, which we can aggregate to ski areas
2. computing the sun's path in the sky throughout the season and adding this to the orientation roses

A couple remaining nice to haves would be:

- a sunlight baseline, for example by computing sunlight for the opposite orientation or equatorward ([brainstorm](https://chatgpt.com/share/67846e4c-4b3c-8011-93cc-aa33265b3d54))
- an visualization, possible interactive, that plots sunlight per segment over the course of a ski day.
